### DISCUSSION TITLE
When to call SSL_shutdown() ?

### DISCUSSION THREAD

**Discussion by rakeshcpp** [openssl/openssl#26205] [Q&A] [answered]
Suppose while receiving response(ssl_read()), I found, in very first chunk(with non-blocking IO) that this is not what I am expecting(some header invalid etc.) then should I immediately close the connection using SSL_shutdown() call, or should I complete reading response first then call SSL_shutdown()?

If I close connection I got this error from ssl just for reference otherwise the above question itself is enough:
`80D0743D4F7D0000:error:0A000123:SSL routines:ssl3_read_bytes:application data after close notify:../ssl/record/rec_layer_s3.c:1660:
`

**Comment by t8m** [comment, marked as answer]
You may call SSL_shutdown() immediately. If you do so, you should either then close the socket immediately (ignoring any remaining data - thus resetting the connection, and without calling SSL_shutdown() a second time). Alternatively you can choose to continue to read the remaining data via SSL_read.
